lib/mayaUsd/ufe/UsdUndoMaterialCommands.cpp

```diff
@@ -375,43 +418,162 @@ void UsdUndoAssignNewMaterialCommand::redo()
+        return nullptr;
+
+    return std::make_shared<UsdUndoAddNewMaterialCommand>(parentItem, nodeId);
+}
+
+Ufe::SceneItem::Ptr UsdUndoAddNewMaterialCommand::insertedChild() const
+{
+    if (_cmds) {
+        auto cmdsIt = _cmds->cmdsList().begin();
+        std::advance(cmdsIt, _createMaterialCmdIdx + 1);
+        auto addShaderCmd = std::dynamic_pointer_cast<UsdUndoCreateFromNodeDefCommand>(*cmdsIt);
+        return addShaderCmd->insertedChild();
+    }
+    return {};
+}
+
+bool UsdUndoAddNewMaterialCommand::CompatiblePrim(const Ufe::SceneItem::Ptr& target)
+{
+    if (!target) {
+        return false;
+    }
+
+    // Must be a scope.
+    if (target->nodeType() != "Scope") {
+        return false;
+    }
+
+    // With the magic name.
+    if (target->nodeName() == UsdUndoAssignNewMaterialCommand::resolvedMaterialScopeName()) {
+        return true;
+    }
+
+    // Or with only materials inside
+    auto scopeHierarchy = Ufe::Hierarchy::hierarchy(target);
+    if (scopeHierarchy) {
+        for (auto&& child : scopeHierarchy->children()) {
+            if (child->nodeType() != "Material") {
+                // At least one non material
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+void UsdUndoAddNewMaterialCommand::execute()
+{
+    if (_parentPath.empty()) {
+        markAsFailed();
+        return;
+    }
+
+    //
+    // Create the Material:
+    //
+    PXR_NS::SdrRegistry&          registry = PXR_NS::SdrRegistry::GetInstance();
     PXR_NS::SdrShaderNodeConstPtr shaderNodeDef
         = registry.GetShaderNodeByIdentifier(TfToken(_nodeId));
-    SdrShaderPropertyConstPtr shaderOutputDef;
-    if (shaderNodeDef->GetSourceType() == "glslfx") {
-        materialOutput = UsdShadeMaterial(materialPrim).CreateSurfaceOutput();
-        shaderOutputDef = shaderNodeDef->GetShaderOutput(TfToken("surface"));
-    } else {
-        if (shaderNodeDef->GetOutputNames().size() != 1) {
-            TF_RUNTIME_ERROR(
-                "Cannot resolve which output of shader %s should be connected to surface",
-                _nodeId.c_str());
-            markAsFailed();
-            return;
-        }
-        materialOutput
-            = UsdShadeMaterial(materialPrim).CreateSurfaceOutput(shaderNodeDef->GetSourceType());
-        shaderOutputDef = shaderNodeDef->GetShaderOutput(shaderNodeDef->GetOutputNames()[0]);
+    if (!shaderNodeDef) {
+        TF_RUNTIME_ERROR("Unknown shader identifier: %s", _nodeId.c_str());
+        markAsFailed();
+        return;
+    }
+    if (shaderNodeDef->GetOutputNames().empty()) {
+        TF_RUNTIME_ERROR("Surface shader %s does not have any outputs", _nodeId.c_str());
+        markAsFailed();
+        return;
+    }
+    auto scopeItem
+        = std::dynamic_pointer_cast<UsdSceneItem>(Ufe::Hierarchy::createItem(_parentPath));
+    auto createMaterialCmd = UsdUndoAddNewPrimCommand::create(
+        scopeItem, shaderNodeDef->GetFamily().GetString(), "Material");
```

Comment:
create() function can return nullptr, so Cmd can be null... are all conditions always guaranteed to be met and never be null?